src/main/java/org/openrewrite/java/testing/dbrider/ExecutionListenerToDbRiderAnnotation.java

```diff
@@ -0,0 +1,275 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * https://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.openrewrite.java.testing.dbrider;
+
+import org.jspecify.annotations.Nullable;
+import org.openrewrite.ExecutionContext;
+import org.openrewrite.Preconditions;
+import org.openrewrite.ScanningRecipe;
+import org.openrewrite.TreeVisitor;
+import org.openrewrite.internal.ListUtils;
+import org.openrewrite.java.AnnotationMatcher;
+import org.openrewrite.java.JavaIsoVisitor;
+import org.openrewrite.java.JavaParser;
+import org.openrewrite.java.JavaTemplate;
+import org.openrewrite.java.search.UsesType;
+import org.openrewrite.java.tree.Expression;
+import org.openrewrite.java.tree.J;
+import org.openrewrite.java.tree.JavaType;
+import org.openrewrite.java.tree.Space;
+
+import java.util.Comparator;
+import java.util.List;
+
+public class ExecutionListenerToDbRiderAnnotation extends ScanningRecipe<ExecutionListenerToDbRiderAnnotation.DbRiderExecutionListenerContext> {
+
+    private static final AnnotationMatcher EXECUTION_LISTENER_ANNOTATION_MATCHER = new AnnotationMatcher("@org.springframework.test.context.TestExecutionListeners");
+    private static final AnnotationMatcher DBRIDER_ANNOTATION_MATCHER = new AnnotationMatcher("@com.github.database.rider.junit5.api.DBRider");
+
+    @Override
+    public String getDisplayName() {
+        return "Migrate the `DBRiderTestExecutionListener` to the `@DBRider` annotation";
+    }
+
+    @Override
+    public String getDescription() {
+        return "Migrate the `DBRiderTestExecutionListener` to the `@DBRider` annotation. " +
+               "This recipe is useful when migrating from JUnit 4 `dbrider-spring` to JUnit 5 `dbrider-junit5`.";
+    }
+
+    @Override
+    public DbRiderExecutionListenerContext getInitialValue(final ExecutionContext ctx) {
+        return new DbRiderExecutionListenerContext();
+    }
+
+    @Override
+    public TreeVisitor<?, ExecutionContext> getScanner(final DbRiderExecutionListenerContext acc) {
+        return new JavaIsoVisitor<ExecutionContext>() {
+            @Override
+            public J.Annotation visitAnnotation(final J.Annotation annotation, final ExecutionContext ctx) {
+                if (EXECUTION_LISTENER_ANNOTATION_MATCHER.matches(annotation)) {
+                    acc.testExecutionListenersFound(annotation);
+                } else if (DBRIDER_ANNOTATION_MATCHER.matches(annotation)) {
+                    acc.dbriderFound = true;
+                }
+                return super.visitAnnotation(annotation, ctx);
+            }
+        };
+    }
```

Comment:
I'm a bit puzzled why this is a scanning recipe, and even more so when this has a boolean that might be set by visiting any of dozens or more test classes, and then affect all of those test classes in turn.

Can we rework this to not be a scanning recipe? You can extract information all the same in a regular visitor by having an inline visitor in the visitCompilationUnit before you call super.visitCompilationUnit for instance. I hope that's clear, or the risk of the current approach is at least clear.

Reply:
Makes total sense... I was thinking of our specific usecase where we typicall have a slice of our tests which have a superclass. But totally ignored the fact that that's absolutely incorrect. Well spotted!

Reply:
Pushed a refactor for this.